### PR TITLE
No shortcut for `narrow` and `meet` in HConsed when int refinement is active

### LINKED ISSUE #1005: Int domain refinement causes fixpoint error
Given the following program:
```c
// PARAM: --set ana.int.refinement fixpoint --enable ana.int.interval
// FIXPOINT
#include<assert.h>

int g = 0;

void main()
{
   int i = 0;
   while (1) {
      i++;
      for (int j=0; j < 10; j++) {
         if (i > 100) g = 1;
      }
      if (i>9) i=0;
   }
   return;
}
```
The verify phase fails with the fixpoint error due to
```
Map: g =
 (Unknown int([0,1]),[0,2147483647]) instead of (Unknown int([0,1]),[0,1]).
```

Although we implement interval refinement from exclusion range, this isn't applied here, probably because int domain refinement is disabled for `join` and `widen`. However, some refinement is required here because the two abstractions have the same concretization, so verification should succeed as it is sound.

There is possible need for int domain refinement in precise incremental benchmarks. Otherwise there are spurious precision comparisons due to similar reasons (exclusion range being more precise than interval).

### PR DESCRIPTION
**Pull request by michael-schwarz**
As @sim642 noted in #1005, int domain refinement (which we do not only after transformers, but also after `meet` and `narrow`), breaks the assumption that `x op x =x` for `narrow` and `meet`.

This disables the shortcut in `HConsed` for `narrow` and `meet` whenever an int refinement is active, and its precondition is not met. It should not be too expensive, as a local boolean to cache this information is introduced.

~Closes #1005.~

(#1005 also mentions a different but related issue that refinement could be applied more often still, but I think this we can save for a separate PR, this PR fixes the immediate issue and makes refinement usable together with hash consing).

### CODE REVIEW

**Comment by sim642** [comment]
This essentially just ignores the issue in the int domains (specifically the reduced product of them) instead of fixing it. I don't think this is the right solution because it still violates the narrowing property $a \sqcap b \sqsubseteq a \mathbin{\Delta} b \sqsubseteq a$. If $a = b$, then this reduces to $a \sqsubseteq a \mathbin{\Delta} a \sqsubseteq a$, which implies $a \mathbin{\Delta} a = a$. The TD3 paper makes this assumption, so we can't just gloss over this.

Before this, `HConsed` presented to TD3 a narrowing which obeys this property. It only failed the additional verification for suddenly becoming more precise. After this, the `narrow` presented to TD3 fails this property, so TD3 might no longer behave as it's supposed to.

What it essentially means is that a stable sequence might become unstable out of nowhere, instead of remaining ultimately stable. Therefore, the TD3 termination proof falls apart.

**Comment by michael-schwarz** [comment]
Ok, we can argue that this should not close the issue then, and the issue should be renamed ("invalid narrowing in the presence of refinement" or something).

However, I still think it is an improvement: The main thing this currently does is make the behavior consistent between having hash-consing on or off. Hash-consing should be a performance optimization only, there should not be any difference observable between the case where it is enabled or disabled. This conflates issues unnecessarily.

The point where your concern applies is immediately when we define the meet and narrow to refine, violating the properties you described above. Hashconsing accidentally fixing that again is not really the way to go.

**Comment by michael-schwarz** [comment]
Relevant paper by Cortesi: https://www.dsi.unive.it/~cortesi/paperi/CL_2011.pdf

**Comment by sim642** [comment]
Actually, `narrow` not being idempotent might not be a problem for termination in practice. It might be in the proof but in practice we don't check if the sequence is _ultimately_ stable. We only check if it's stable for one iteration and end narrowing there. Unlike widening, it's perfectly fine to end narrowing at any point. It doesn't even have to reach one step stability. We can just get a less precise result maybe.

**Comment by michael-schwarz** [comment]
Merging this as per discussion at the GobCon yesterday.

File: src/common/domains/lattice.ml [hunk 1]
```diff
@@ -151,10 +151,15 @@ end
 module HConsed (Base:S) =
 struct
   include Printable.HConsed (Base)
+
+  (* We do refine int values on narrow and meet {!IntDomain.IntDomTupleImpl}, which can lead to fixpoint issues if we assume x op x = x *)
+  (* see https://github.com/goblint/analyzer/issues/1005 *)
+  let int_refine_active = GobConfig.get_string "ana.int.refinement" <> "never"
```
[the remark this answers is not in the document]
Comment (michael-schwarz) [reply src/common/domains/lattice.ml:157]: This module is constructed dynamically though, so it is not really a problem here.

**Pull request merged**